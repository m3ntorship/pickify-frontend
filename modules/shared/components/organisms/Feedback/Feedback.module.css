.text {
  @apply text-dark-grey text-sm leading-5 font-normal;
}

.sub-text {
  @apply text-dark-grey text-xs leading-4 font-light;
}

.emojis {
  @apply flex flex-row mb-4 mt-2;
}

.button {
  @apply flex justify-center mt-3;
}

.submitted-container {
  @apply flex flex-col justify-center items-center whitespace-nowrap px-4xxl py-4xl;
}

.form-answer {
  @apply inline-block;
}

.emoji {
  @apply opacity-80 w-full block box-border mr-4 cursor-pointer
  filter grayscale;
}

.form-answer input[type='radio'] {
  @apply w-0 h-0 opacity-0 hidden;
}

.form-answer input[type='radio']:checked ~ label,
.form-answer input[type='radio']:hover ~ label {
  @apply filter-none opacity-100;
}

.text-area {
  @apply border border-solid border-grey-shd5 rounded-md
   placeholder-grey-shd1
   pt-xsvv pl-4 mt-2
   focus:outline-none  focus:border-grey-shd5
   ;
}
