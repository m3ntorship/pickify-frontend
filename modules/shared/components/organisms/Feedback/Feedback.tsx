import React, { useState } from 'react';
import type { FC, ReactElement } from 'react';
import styles from './Feedback.module.css';
import Button from '../../atoms/Button/Button';
import * as EButton from '../../atoms/Button/types/EButton';
import Terrible from '../../icons/terrible.svg';
import Good from '../../icons/good.svg';
import Bad from '../../icons/bad.svg';
import Neutral from '../../icons/neutral.svg';
import Amazing from '../../icons/amazing.svg';
import TextInput from '../../atoms/TextInputs/TextInput';
import * as ETextInput from '../../atoms/TextInputs/types/ETextInput';

const Feedback: FC = (): ReactElement => {
  const emojis: { name: string; checked: boolean; component: JSX.Element }[] = [
    {
      name: 'terrible',
      checked: false,
      component: <Terrible />,
    },
    {
      name: 'bad',
      checked: false,
      component: <Bad />,
    },
    {
      name: 'neutral',
      checked: false,
      component: <Neutral />,
    },
    {
      name: 'good',
      checked: false,
      component: <Good />,
    },
    {
      name: 'amazing',
      checked: false,
      component: <Amazing />,
    },
  ];
  const [checkedValue, setCheckedValue] = useState(''); // set default value here to be defaultChecked or put ''
  const [disabled, setDisabled] = useState(true);
  const [inputValue, setInputValue] = useState('');
  const [isSubmitted, setIsSubmitted] = useState(false);

  const onChangeInputValueHandler = (
    id: string,
    e: React.ChangeEvent<HTMLInputElement>,
  ): void => {
    setInputValue(e.target.value);
  };
  const onClickDeleteInputValueHandler = (): void => {
    setInputValue('');
  };

  const positiveOrNegativeFeedback = (): string => {
    if (checkedValue === 'good' || checkedValue === 'amazing') {
      return 'positiveFeedback';
    }
    return 'negativeFeedback';
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setCheckedValue(e.target.value);
    setDisabled(false);
  };

  const handleSubmit = (): void => {
    setIsSubmitted(true);
  };
  return (
    <div className={styles.container}>
      {!isSubmitted && (
        <form onSubmit={handleSubmit}>
          <h1 className={styles.text}>How would you rate your experience?</h1>
          <div className={styles.emojis}>
            {emojis.map((emoji) => {
              return (
                <>
                  <fieldset className={styles['form-options']}>
                    <p className={styles['form-answer']}>
                      <input
                        type="radio"
                        onChange={handleChange}
                        name="emoji"
                        id={emoji.name}
                        value={emoji.name}
                      />
                      <label className={styles.emoji} htmlFor={emoji.name}>
                        {emoji.component}
                      </label>
                    </p>
                  </fieldset>
                </>
              );
            })}
          </div>

          <div>
            <TextInput
              label="Do you have any comments?"
              id="my label"
              inputType={ETextInput.InputType.Default}
              variants={ETextInput.Variants.Default}
              disabled={false}
              value={inputValue}
              placeholder="Enter your feedback"
              onChangeInputValueHandler={onChangeInputValueHandler}
              onClickDeleteInputValueHandler={onClickDeleteInputValueHandler}
              onBlurInputHandler={(): boolean => true}
            />
          </div>

          <div className={styles.button}>
            <Button
              size={EButton.buttonSizeValues.XLARGE}
              variant={EButton.buttonVariantValues.PRIMARY}
              disabled={disabled}
              buttonText="Submit"
              buttonType="submit"
            />
          </div>
        </form>
      )}

      {isSubmitted && positiveOrNegativeFeedback() === 'negativeFeedback' && (
        <div className={styles['submitted-container']}>
          <Terrible className="mb-6" />
          <h1 className={styles.text}>Sorry to hear about your experience!</h1>
          <h4 className={styles['sub-text']}>
            We will do our best to improve it
          </h4>
        </div>
      )}
      {isSubmitted && positiveOrNegativeFeedback() === 'positiveFeedback' && (
        <div className={styles['submitted-container']}>
          <Amazing className="mb-6" />
          <h1 className={styles.text}>Thanks for your feedback!</h1>
        </div>
      )}
    </div>
  );
};
export default Feedback;
