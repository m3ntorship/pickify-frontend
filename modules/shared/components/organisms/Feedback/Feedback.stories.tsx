import type { ReactElement } from 'react';
import type { Story } from '@storybook/react';
import Feedback from './Feedback';

export default {
  title: 'organisms/Feedback',
  component: Feedback,
};

const Template: Story = (args): ReactElement => <Feedback {...args} />;
export const Default = Template.bind({});
Default.args = {};
