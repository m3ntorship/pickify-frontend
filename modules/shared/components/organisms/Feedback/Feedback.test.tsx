import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Feedback from './Feedback';

describe('Feedback', () => {
  it('renders The Feedback initial state', () => {
    const tree = renderer.create(<Feedback />);
    expect(tree).toMatchSnapshot();
  });

  it('renders The Feedback with disabled button as default', () => {
    render(<Feedback />);

    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();
  });

  it('renders The Feedback with disabled button as default and when user chose any emoji it will be activated', () => {
    render(<Feedback />);
    const goodEmoji = screen.getByTestId('good');
    const submitButton = screen.getByRole('button');

    userEvent.click(goodEmoji);

    expect(submitButton).toBeEnabled();
  });

  it('renders The Feedback when user click happy or amazing emoji state', () => {
    render(<Feedback />);

    const goodEmoji = screen.getByTestId('good');
    const submitButton = screen.getByRole('button');

    userEvent.click(goodEmoji);
    userEvent.click(submitButton);

    const positiveFeedback = screen.getByTestId('positiveFeedback');

    expect(positiveFeedback).toBeInTheDocument();
  });
  it('renders The Feedback when user click bad or neutral or terrible emoji state', () => {
    render(<Feedback />);

    const badEmoji = screen.getByTestId('bad');
    const submitButton = screen.getByRole('button');

    userEvent.click(badEmoji);
    userEvent.click(submitButton);

    const negativeFeedback = screen.getByTestId('negativeFeedback');

    expect(negativeFeedback).toBeInTheDocument();
  });
});
