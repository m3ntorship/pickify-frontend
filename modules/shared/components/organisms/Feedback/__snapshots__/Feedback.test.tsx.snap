// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback renders The Feedback initial state 1`] = `
<div
  className="box bg-white"
>
  <div
    className=""
  >
    <form
      onSubmit={[Function]}
    >
      <h1
        className="text"
      >
        How would you rate your experience?
      </h1>
      <div
        className="emojis"
      >
        <fieldset
          className="form-options"
        >
          <p
            className="form-answer"
          >
            <input
              id="terrible"
              name="emoji"
              onChange={[Function]}
              type="radio"
              value="terrible"
            />
            <label
              className="emoji"
              data-testid="terrible"
              htmlFor="terrible"
            >
              <svg
                data-file-name="SvgTerrible"
              />
            </label>
          </p>
        </fieldset>
        <fieldset
          className="form-options"
        >
          <p
            className="form-answer"
          >
            <input
              id="bad"
              name="emoji"
              onChange={[Function]}
              type="radio"
              value="bad"
            />
            <label
              className="emoji"
              data-testid="bad"
              htmlFor="bad"
            >
              <svg
                data-file-name="SvgBad"
              />
            </label>
          </p>
        </fieldset>
        <fieldset
          className="form-options"
        >
          <p
            className="form-answer"
          >
            <input
              id="neutral"
              name="emoji"
              onChange={[Function]}
              type="radio"
              value="neutral"
            />
            <label
              className="emoji"
              data-testid="neutral"
              htmlFor="neutral"
            >
              <svg
                data-file-name="SvgNeutral"
              />
            </label>
          </p>
        </fieldset>
        <fieldset
          className="form-options"
        >
          <p
            className="form-answer"
          >
            <input
              id="good"
              name="emoji"
              onChange={[Function]}
              type="radio"
              value="good"
            />
            <label
              className="emoji"
              data-testid="good"
              htmlFor="good"
            >
              <svg
                data-file-name="SvgGood"
              />
            </label>
          </p>
        </fieldset>
        <fieldset
          className="form-options"
        >
          <p
            className="form-answer"
          >
            <input
              id="amazing"
              name="emoji"
              onChange={[Function]}
              type="radio"
              value="amazing"
            />
            <label
              className="emoji"
              data-testid="amazing"
              htmlFor="amazing"
            >
              <svg
                data-file-name="SvgAmazing"
              />
            </label>
          </p>
        </fieldset>
      </div>
      <div>
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="my label"
          >
            Do you have any comments?
          </label>
          <div
            className="form-control"
          >
            <input
              className="form-input"
              data-input-type="default"
              data-testid="text-input"
              data-variant="default"
              dir="auto"
              disabled={false}
              id="my label"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Enter your feedback"
              type="text"
              value=""
            />
            <span
              className="status-icon"
            >
              
            </span>
            <span
              className="control-icon"
            />
          </div>
        </div>
      </div>
      <div
        className="button"
      >
        <button
          className="btn btn-primary btn-xlarge"
          disabled={true}
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
